src/DDMinimizer.cpp

```diff
@@ -0,0 +1,445 @@
+  // to index
+  std::map<pair<Qubit, Qubit>, int> xCMap;
+  std::map<pair<Qubit, Qubit>, int> cXMap;
+  for (size_t i = 0; i < max; i++) {
+    xCMap.insert({{i + 1, i}, -1});
+    cXMap.insert({{i, i + 1}, -1});
+  }
+  // save complete maps of ladder and for max index evaluation
+  maps.insert({"xC", xCMap});
+  indices.insert({"xC", std::vector<int>(1, 0)});
+  maps.insert({"cX", cXMap});
+  indices.insert({"cX", std::vector<int>(1, 0)});
+
+  // create c-l and x-l ladder
+  for (size_t i = 0; i < max; i++) {
+    std::map<pair<Qubit, Qubit>, int> xLMap;
+    std::map<pair<Qubit, Qubit>, int> cLMap;
+    // create the steps of the ladder
+    for (size_t j = 0; j < bits; j++) {
+      if (i < j) {
+        xLMap.insert({{j, i}, -1});
+        cLMap.insert({{i, j}, -1});
+      }
+    }
+    // save the steps in the maps
+    maps.insert({"xL" + std::to_string(i), xLMap});
+    maps.insert({"cL" + std::to_string(i), cLMap});
+  }
+  // save the complete ladder for max index evaluation
+  indices.insert({"xL", std::vector<int>(bits - 1, 0)});
+  indices.insert({"cL", std::vector<int>(bits - 1, 0)});
+
+  // create c-r and x-r ladder
+  for (size_t i = max; i > 0; i--) {
+    std::map<pair<Qubit, Qubit>, int> xRMap;
+    std::map<pair<Qubit, Qubit>, int> cRMap;
+    // create the steps of the ladder
+    for (size_t j = 0; j < bits; j++) {
+      if (i > j) {
+        xRMap.insert({{j, i}, -1});
+        cRMap.insert({{i, j}, -1});
+      }
+    }
+    // save the steps in the maps
+    maps.insert({"xR" + std::to_string(i), xRMap});
+    maps.insert({"cR" + std::to_string(i), cRMap});
+  }
+  // save the complete ladder for max index evaluation
+  indices.insert({"xR", std::vector<int>(bits - 1, 0)});
+  indices.insert({"cR", std::vector<int>(bits - 1, 0)});
+
+  return std::make_pair(maps, indices);
+}
+
+// Helper function to check if the vector of a ladder step is full, meaning each
+// gate appeared in the circuit
+bool DDMinimizer::isFull(const std::vector<int>& vec) {
+  return std::all_of(vec.begin(), vec.end(),
+                     [](int value) { return value != -1; });
+}
+
+// Helper function to get the number of complete stairs in a ladder
+int DDMinimizer::getStairCount(const std::vector<int>& vec) {
+  int count = 0;
+  for (const int value : vec) {
+    if (value != -1) {
+      count++;
+    } else {
+      return count;
+    }
+  }
+  return count;
+}
+
+// Helper function to get the position of a ladder in comparison to other steps
+int DDMinimizer::getLadderPosition(const std::vector<int>& vec, int ladder) {
+  int count = 0;
+  for (size_t i = 0; i < vec.size(); i++) {
+    const int value = vec[i];
+
+    if (i == vec.size() - 1 && count != 0) {
+      if (value != -1 && value < ladder) {
+        count++;
+      }
+    }
+  }
+  return count;
+}
+
+// Helper functions to manipulate the Layout: reverse, rotate right and left
+
+std::vector<Qubit> DDMinimizer::reverseLayout(std::vector<Qubit> layout) {
```

Comment:
When you do not declare the parameter as a reference, that the parameter is passed by value which is an additional copy operation.
```suggestion
std::vector<Qubit> DDMinimizer::reverseLayout(std::vector<Qubit>& layout) {
```